src/main/java/at/hannibal2/skyhanni/data/IslandGraphs.kt

```diff
@@ -260,64 +296,185 @@ object IslandGraphs {
         setFastestPath(path to (distance + nodeDistance))
     }
 
-    private fun setFastestPath(path: Pair<Graph, Double>) {
-        fastestPath = path.takeIf { it.first.isNotEmpty() }?.first
+    private fun Graph.totalLenght(): Double = nodes.zipWithNext().sumOf { (a, b) -> a.position.distance(b.position) }
+
+    private fun handlePositionChange() {
+        val secondClosestNode = secondClosedNote ?: return
+        distanceToNextNode = secondClosestNode.position.distanceToPlayer()
+        updateChat()
+    }
+
+    private var hasMoved = false
 
-        fastestPath?.let {
-            fastestPath = Graph(cutByMaxDistance(it.nodes, 3.0))
+    private fun checkMoved() {
+        if (hasMoved) {
+            hasMoved = false
+            if (goal != null) {
+                handlePositionChange()
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onPlayerMove(event: EntityMoveEvent) {
+        if (LorenzUtils.inSkyBlock && event.entity == Minecraft.getMinecraft().thePlayer) {
+            hasMoved = true
+        }
+    }
+
+    private fun setFastestPath(path: Pair<Graph, Double>, setPath: Boolean = true) {
+        val (fastestPath, distance) = path.takeIf { it.first.isNotEmpty() } ?: return
+        val nodes = fastestPath.nodes.toMutableList()
+        if (Minecraft.getMinecraft().thePlayer.onGround) {
+            nodes.add(0, GraphNode(0, LocationUtils.playerLocation()))
+        }
+        if (setPath) {
+            this.fastestPath = Graph(cutByMaxDistance(nodes, 3.0))
         }
+
+        val diff = fastestPath.getOrNull(1)?.let {
+            fastestPath.first().position.distance(it.position)
+        } ?: 0.0
+        this.distanceViaNodes = distance - diff
+        updateChat()
     }
 
     private fun onNewNote() {
         // TODO create an event
         IslandAreas.noteMoved()
+        if (shouldAllowRerouting) {
+            tryRerouting()
+        }
+    }
+
+    private fun tryRerouting() {
+        val target = currentTargetNode ?: return
+        val closest = closedNote ?: return
+        val graph = currentIslandGraph ?: return
+        val sameNodes = mutableMapOf<GraphNode, Double>()
+        for (node in graph.nodes) {
+            if (target.sameNameAndTags(node)) {
+                val (_, distance) = graph.findShortestPathAsGraphWithDistance(closest, node)
+                sameNodes[node] = distance
+            }
+        }
```

Comment:
nea made a function that returns all distance from that graphnode to any other us that instead. Also make the sameNameAndTags a lookup instead of a iteration over the whole graph.